EnumWidener

nnym opened the pull request:
A **compilation time-only** JAR processor that unsets the `enum` modifier from enums and their fields and fixes enum constructor signatures to make name and ordinal parameters accessible. Further, it makes the synthetic enum array visible. Combined with AccessWidener, it allows enums to be instantiated (via the `new` opcode; the JLS does not forbid it) and extended.

Since unsetting the `enum` modifier at runtime is dangerous, compilation and runtime dependencies have to be separated. This separation is achieved by replacing the previous `minecraftNamed` configuration by 2 new configurations: `minecraftNamedCompile` and `minecraftNamedRuntime` that correspond to separate JAR dependencies, and adding appropriate archive classifiers in order to differentiate them.

usage in `build.gradle`: 
```groovy
minecraft {
	enumWidener("net/minecraft/SomeEnum")
}
```

nnym committed:
EnumWidener™ v0

nnym committed:
split JAR into compilation and runtime

nnym committed:
save hash per JAR

nnym committed:
Merge branch 'dev/0.5' of github.com:FabricMC/fabric-loom into enum-widener
pull upstream

nnym committed:
yes

nnym committed:
fix compilation errors due to mismatch in constructor parameter annotation count and new signature length

nnym committed:
license header

nnym committed:
retain method synthetic modifier

nnym committed:
final is crabbed

nnym committed:
restore build.gradle

nnym committed:
fix license padding

nnym committed:
fix Checkstyle

ramidzkh commented on src/main/java/net/fabricmc/loom/LoomGradleExtension.java:
```diff
@@ -151,6 +152,14 @@ public JsonObject getInstallerJson() {
 
+	public void enumWidener(String klass) {
+		if (this.enumWidener == null) {
+			this.enumWidener = new ArrayList<>();
```
Maybe initialize the field in the constructor instead? Would look neater

ramidzkh commented on src/main/java/net/fabricmc/loom/processors/EnumWidenerJarProcessor.java:
```diff
@@ -0,0 +1,84 @@
+						.toArray(ZipEntryTransformerEntry[]::new)
+			);
+
+			ZipUtil.addEntry(file, HASH_FILE_NAME, ByteBuffer.allocate(4).putInt(this.classes.hashCode()).array());
```
Access wideners do a SHA-256 hash instead

nnym replied:
Should I also do that? Is something wrong with `hashCode`?

ramidzkh commented on src/main/java/net/fabricmc/loom/processors/EnumWidenerJarProcessor.java:
```diff
@@ -0,0 +1,84 @@
+
+	@Override
+	public boolean isInvalid(File file) {
+		if (this.classes == null || !file.equals(this.getTarget())) {
```
~~If the enumwideners were removed, the file would be marked not invalid.~~ Seems to be a flaw with the current design of the current JarProcessor system really, as the same is true if all access wideners are removed but another JarProcessor is present.

~~I would also remove the getTarget methods and its associated checks~~ Maybe make JarProcessors aware if they are operating on the compile or runtime jar?

ramidzkh commented on src/main/java/net/fabricmc/loom/processors/MinecraftProcessedProvider.java:
```diff
@@ -50,23 +53,27 @@ public MinecraftProcessedProvider(Project project, JarProcessorManager jarProces
 	@Override
 	protected void addDependencies(DependencyInfo dependency, Consumer<Runnable> postPopulationScheduler) {
-		if (jarProcessorManager.isInvalid(projectMappedJar) || isRefreshDeps()) {
+		if (jarProcessorManager.isInvalid(projectMappedCompileJar) || isRefreshDeps()) {
```
Check if both are invalid

asiekierka commented:
How does this interact with enum switch statements? Particularly, a case with a non-void method which has its return statements in an enum switch-case; does this tackle that problem, or is it left to the programmer to avoid such cases?

i509VCB commented on src/main/java/net/fabricmc/loom/util/enumwidener/EnumWidenerClassNode.java:
```diff
@@ -0,0 +1,63 @@
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.ClassNode;
+
+public class EnumWidenerClassNode extends ClassNode {
```
Class visitor?

Ueaj-Kerman replied:
no, I think he's trying to save on object here. kindof a microoptimization but whatever

i509VCB replied:
You'd still have to allocate the classreader and classwriter though. Just the visitor would be lower overhead.

nnym replied:
I had a ClassVisitor earlier but I needed a ClassNode in order to shift the parameter annotations due to the 2 additional parameters in constructor signatures.

Chocohead replied:
What was wrong with just returning a `MethodVisitor` which added two to `visitParameterAnnotation`'s `parameter`s when the method is `<init>`?

nnym replied:
> What was wrong with just returning a `MethodVisitor` which added two to `visitParameterAnnotation`'s `parameter`s when the method is `<init>`?

Nothing; I had not thought that it would be possible. It should be better now.

Ueaj-Kerman commented:
https://github.com/SpongePowered/Mixin/issues/387

According to 🌍🖥️ it's not a problem. no idea how it works, but it works, and if oracle says adding enums doesn't break binary compatibility, you best believe it doesn't. oracle will break your legs if you break binary compat

Ueaj-Kerman reviewed (approved):
if it works, looks good

i509VCB commented:
Wouldn't this need a runtime component? or does it not need that.

Arc-blroth commented:
As far as I know, this would require a runtime component - access hasn't added that yet as far as I know

modmuss50 commented on src/main/java/net/fabricmc/loom/AbstractPlugin.java:
```diff
@@ -115,8 +115,10 @@ public void apply(Project target) {
 
-		Configuration minecraftNamedConfig = project.getConfigurations().maybeCreate(Constants.Configurations.MINECRAFT_NAMED);
-		minecraftNamedConfig.setTransitive(false); // The launchers do not recurse dependencies
+		Configuration minecraftNamedCompileConfig = project.getConfigurations().maybeCreate(Constants.Configurations.MINECRAFT_NAMED_COMPILE);
```
It should only use the split configurations if required.

Arc-blroth commented:
> the JLS does not forbid it

Actually, re-reading access' original post, this would not require a runtime component

modmuss50 commented on src/main/java/net/fabricmc/loom/processors/MinecraftProcessedProvider.java:
```diff
@@ -50,23 +53,27 @@ public MinecraftProcessedProvider(Project project, JarProcessorManager jarProces
 
-		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED,
-				getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(PROJECT_MAPPED_CLASSIFIER)));
+		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED_COMPILE,
+				getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(PROJECT_MAPPED_COMPILE_CLASSIFIER)));
+		getProject().getDependencies().add(Constants.Configurations.MINECRAFT_NAMED_RUNTIME,
+					getProject().getDependencies().module("net.minecraft:minecraft:" + getJarVersionString(PROJECT_MAPPED_RUNTIME_CLASSIFIER)));
```
Again here, only use the split jar if required.

modmuss50 commented on src/main/java/net/fabricmc/loom/processors/MinecraftProcessedProvider.java:
```diff
@@ -50,23 +53,27 @@ public MinecraftProcessedProvider(Project project, JarProcessorManager jarProces
 			}
 
-			jarProcessorManager.process(projectMappedJar);
+			jarProcessorManager.process(projectMappedCompileJar);
+			jarProcessorManager.process(projectMappedRuntimeJar);
```
No need to run the same processors on 2 jars.

I think the jar processor API should get a bit of an overhall to support this in a much nicer way.

i509VCB commented:
> enumWidener("net/minecraft/SomeEnum")

Should every enum have to be specified in loom or do we want to do something like a new file or make it part of AW and just ignore it at runtime.

Ueaj-Kerman commented:
I don't think you'll be widening enough enums to justify a separate file

dexman545 commented:
How are `default` cases of enum switcghes handled, particularly the ones that throw an exception? Should they just be avoided? If so, in a multimod environment that could lead to issues - some mod adds another `Hand`, but another mod decided to throw on any unexpected `Hand` value.

i509VCB commented:
For exceptions you could just inject at NEW and cancel? However for cases where there is more logic afterwards that is not suitable.

Whats does jvm do if you try to throw null?

nnym commented:
> How does this interact with enum switch statements? Particularly, a case with a non-void method which has its return statements in an enum switch-case; does this tackle that problem, or is it left to the programmer to avoid such cases?

This pull request does not tackle that problem.

> > enumWidener("net/minecraft/SomeEnum")
> 
> Should every enum have to be specified in loom or do we want to do something like a new file or make it part of AW and just ignore it at runtime.

I considered using AccessWidener, but as it currently is, adding a single transformation could require the addition of multiple (up to 4?) `ClassAccess` constants. Besides that, EnumWidener has to do additional transformations to fix constructor signatures. Since it exists only during compilation, I think that an entry in the build script for each enum should suffice.

modmuss50 reviewed (commented):
As for the loom side of it, I do see the use of having a split compile/runtime jar but it needs to be a bit more inteligent than this. (Only using 2 jars when required, and only applying processors when needed not twice). - Should be seprate PR IMO, when I have some more time I can possibly write down my thoughts on this, or even do the work.

As for the enum widener (needs another name) bit of it, I like the idea; but another possible apporch would be to have a data driven way of adding enum entries (would remove the need of the split jars) Most of the time enum values are constructed via basic to no inputs (strings, ints and what not) these could be quite easily defined in a custom format. Failing that it could be pointed to a static method in another class.

```
net/minecraft/ExampleTargetEnum
   VALUE
      1
      "A string"
      my/mod/EnumUtils.aStaticMethod
```

This is just what ive come up with now as a basic example of what this could look like. The int, string and method call would be passed into the enums constructor. This would get remapped to intermediary on build.

As with access widners I dont think the logic should go into loom but into another lib but thats something thats not an issue at this stage.

The final question is, does this fit into fabric by default? Im not 100% sure it does, this wil need discussion and evluation of what others thing. I am happy to look into expanding the loom API to allow other plugins to have greater control over how things work, however I imagine loader's (lack of) API around this area will be the biggest pain point.

nnym committed:
use ClassVisitor and MethodVisitor instead of ClassNode and MethodNode

nnym committed:
license header

nnym committed:
remove final

nnym committed:
check both JARs for validity

nnym committed:
fix invalid check and initialize enumWidener in field declaration

Chocohead commented on src/main/java/net/fabricmc/loom/util/enumwidener/EnumWidenerClassVisitor.java:
```diff
@@ -0,0 +1,63 @@
+
+	@Override
+	public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+		return new EnumWidenerMethodVisitor(
+			Constants.ASM_VERSION,
+			super.visitMethod(access, name, descriptor, name.equals("<init>") ? signature.replace("(", "(Ljava/lang/String;I") : signature, exceptions),
+			name.equals("<init>")
+		);
+	}
```
You might as well only use the `EnumWidenerMethodVisitor` for `<init>` as there's optimisations within `ClassReader` to pass straight to a `MethodWriter` (i.e. if there's no intermediary `MethodVisitor`s)

Chocohead commented on src/main/java/net/fabricmc/loom/util/enumwidener/EnumWidenerTransformerEntry.java:
```diff
@@ -0,0 +1,57 @@
+
+	@Override
+	protected byte[] transform(ZipEntry zipEntry, byte[] input) {
+		ClassWriter writer = new ClassWriter(0);
```
If you keep the `ClassReader` then pass it to this as a constructor argument it can reuse the symbol table rather than have to remake it

nnym committed:
fix IJ IDEA formatting

nnym committed:
split JARs intelligently

nnym committed:
adapt to split JAR API

nnym committed:
Environment license header

nnym committed:
environment awareness

nnym committed:
fix conflict

nnym committed:
fix imports

nnym committed:
fix imports

nnym committed:
fix crash when signature is null

nnym committed:
add EnumWidenerJarProcessor if classes are specified

modmuss50 committed:
Start 0.6 dev cycle

modmuss50 committed:
Move remapped mods into a full local maven repo (#300)

modmuss50 committed:
Refactor and fix game/loader library handling (#298)

Update fernflower

modmuss50 committed:
Use launcher metadata manifest v2. Closes #290

nnym committed:
Add nogui to server program arguments (#310)

Closes #276

CoolMineman committed:
Respect reproducibleFileOrder and preserveFileTimestamps (#304)

* Respect reproducibleFileOrder and preserveFileTimestamps

* Also Respect On Sources Jars

* java 8

* checkstyle

* Please Rerun Actions

solonovamax committed:
Fix genSources not working when root project doesnt have loom applied#308. (#309)

loom now runs genSources on root project if the plugin exists in the root project, else it will build in the current project.

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>

modmuss50 commented on .editorconfig:
```diff
@@ -1,2 +1,4 @@
 [*.{gradle,java}]
 indent_style = tab
+ij_continuation_indent_size = 8
+ij_java_imports_layout = $*,|,java.**,|,*,|,net.fabricmc.**
```
Going to steel this and put it in on its own. Thanks.

modmuss50 committed:
Update editorconfig to set import order, thanks to @user11681

modmuss50 committed:
Allow setting the RemapJarTask classpath. Closes #307

i509VCB committed:
Add ability to specify additional tiny remapper options in remapJar (#292)

* Add ability to specify additional tiny remapper options in remapJar

* Imports go brr

* Fix checkstyle

Co-authored-by: modmuss50 <modmuss50@gmail.com>

modmuss50 committed:
General code cleanup (#313)

* First general cleanup pass

* Review feedback

* Fix build

* Fix tests

modmuss50 committed:
Remove jcenter as a default maven repo. This will make builds faster and more reliable. It is trivial to add back manually for projects that require it.

Closes #314

modmuss50 committed:
Optimise loom configuration, saves 1.2 seconds in my testing. #319

modmuss50 committed:
Fix classifier handling for remapped dependencies

modmuss50 committed:
Cleanup output files when remap fails. Fixes #321

modmuss50 committed:
Improve the readme a little.

Feel free to improve via PRs

Juuxel committed:
Improve debugging repo instructions and fix typos (#323)

* Improve debugging repo instructions and fix typos

* Fix typo in README

modmuss50 committed:
Remove usage of internal kotlin class, should fix crash with kotlin 1.4. Closes #324

nnym committed:
pull dev/0.6

modmuss50 committed:
Update a number of dependencies

modmuss50 committed:
Fix build.

Thanks gradle for changing versioning style..

modmuss50 committed:
Remove dep on procyon-fabric-compilertools I cannot see its use

modmuss50 committed:
Gradle 7 (and hopefully 8) support (#332)

* First pass on gradle 7 support

* Fix Gradle 8 deprecation warnings

* Add a deprecated compile constant

modmuss50 committed:
1 Infinite Loop

modmuss50 committed:
Centralise refreshDeps variable.
Makes it easier to hack it to always enable for profiling.

modmuss50 committed:
Dont remap dependency sources when building on CI services.
Use `-Dfabric.loom.ci=false` to force disable

modmuss50 committed:
Apply all buildscript configurations across all parent projects to try and fix fernflower classpath issues again.

If you have any better suggestions please let me know.

modmuss50 committed:
Another attempt at fixing issues with fernflower's classpath

modmuss50 committed:
Improve error handling around access widener remapping. Closes #311

modmuss50 committed:
Revert previous attempt at fixing genSources when loom is applied via buildSrc

nnym committed:
Merge branch 'dev/0.6' of github.com:FabricMC/fabric-loom into enum-widener

nnym committed:
pull dev/0.6

nnym committed:
use compilation dependencies for AP

nnym committed:
move Environment to the correct package

nnym committed:
pull split-jar

Chocohead committed:
Use the given manifest version hash if possible (#337)

bytzo committed:
Set classpath before compileJava task (#330) (#333)

modmuss50 committed:
Dont add the mojang maven repo twice.

Might speed some things up as well due to ordering

modmuss50 committed:
Merge remote-tracking branch 'FabricMC/dev/0.6' into dev/0.6

shedaniel committed:
Cache the checksum of asset index files. (#335)

* Cache the checksum of asset index files.

* Resolve reviews

modmuss50 committed:
Improve run configs (#318) (#338)

Co-authored-by: ShadewRG <onankman@gmail.com>
Co-authored-by: UndeadCat | Shadew <redgalaxysw@gmail.com>
Co-authored-by: liach <7806504+liach@users.noreply.github.com>

modmuss50 committed:
Rename mode -> environment + Hopefully fix a regression with the run tasks

jpenilla committed:
Generate fabric.mod.json for included submodule jars (#343)

* Generate fabric.mod.json for included submodule jars

* Appease checkstyle

modmuss50 committed:
Fix CME, Closes #345

modmuss50 committed:
Update MCE - Remove auto population of minVersion in mixins - Drop some log levels.

modmuss50 committed:
Rename `runs` -> `runConfigs`

modmuss50 committed:
Revert previous change, somewhat. I didnt understand the issue correctly.

modmuss50 committed:
Generate IDE run configs on a per config basis. Allows for much finer grain control over what run configs are generated.

Defaults to generate all run configs on the root project.

modmuss50 committed:
Rewrite minecraft game data handling.

Much faster asset and native validation logic.
Simplfied game metadata model.

modmuss50 committed:
asm 9.1

Juuxel committed:
RunConfigSettings.generateIDEConfig -> ideConfigGenerated + normal getter/setter (#349)

Juuxel committed:
Add dev runtime dependencies to runtimeClasspath instead of runtimeOnly (#350)

This stops them from incorrectly showing up in Gradle module metadata
when a software component is used.

modmuss50 committed:
Disable log4j jansi support when compiling, removes a warning about it.

modmuss50 committed:
Run tests against gradle 7 and fail on deprecation. Fix gradle 8 support again.

modmuss50 committed:
Add loom dev deps to runtimeClasspath instead of runtimeOnly

shadowfacts committed:
Add log4jConfigs to Loom extension (#356)

Allows mod build scripts to specify custom Log4j config files that will
be combined together when running Minecraft in the dev-env. For example:

    loom {
        log4jConfigs.from "MyCustomConfig.xml"
    }

See: https://logging.apache.org/log4j/2.x/manual/configuration.html#CompositeConfiguration

modmuss50 committed:
Fix inherit not copying `ideConfigGenerated`

modmuss50 committed:
Start 0.7 dev cycle

modmuss50 committed:
Fix custom sourceset not being correctly set for runconfig tasks (#357)

modmuss50 committed:
Minor cleanup/optimisations to downloading, should help a little bit more with asset downloading.
Closes #359

modmuss50 committed:
Rewrite tests to be easier to write and maintain (#361)

modmuss50 committed:
Fix ReproducibleBuildTest

modmuss50 committed:
Try and fix fernflower forking for the final time. Hopefully helps with #355

modmuss50 committed:
Increase heapsize for tests

modmuss50 committed:
Fix tests on windows

modmuss50 committed:
Add basic MultiProjectTest and a basic RunConfigTest

This should cover most of the main things now.

modmuss50 committed:
Mod Remapping Refactor. (#362)

Improved gradle 7 support
General cleanup

modmuss50 committed:
Ensure outputs are reproducable across all OS's. (#363)

modmuss50 committed:
Fix case where it wont try to download file.

modmuss50 committed:
Run tests against 7.0-rc-1

modmuss50 committed:
Revert some of the changes to genSource's.

More work is needed to fix all the issues with it but its not really something I want to get into right now.

modmuss50 committed:
Constant unpicking (#328)

* Start adding constant unpicking

* Update to use unpick cli

* Fix build?

* Fix?

* Fix log spam when unpicking

* Improve unpick tests

modmuss50 committed:
Add basic groovy code style validation.

IDK how well this would work on a gradle script but might be interesting to try on yarn or something.

modmuss50 committed:
Record javadoc support. (#366)

* Record javadoc support

* Ensure we dont add JD twice

modmuss50 committed:
Add MavenProjectTest (#372)

octylFractal committed:
Fix name comparision in GroovyXmlUtil (#373)

Nodes can also have groovy.xml.QNames, which need to be compared using
their matches(Object) method.

modmuss50 committed:
Ensure output is deleted before remapping the game.

modmuss50 committed:
Fix last commit, oops

modmuss50 committed:
Reorganise tests (#378)

modmuss50 committed:
Update UnpickTest to latest snapshot.

This ensures that constants included in the mc jar can be used.

modmuss50 committed:
Support Groovy 3's QName in GroovyXmlUtil (#379)

modmuss50 committed:
Update to Gradle 7 (#380)

* Gradle 7 part 1

* Fix actions?

* Update actions

* Fix tests not running

* Fix some broken tests

* Update kotlin to try and fix JDK 16 support

* Disable KotlinTest on j16

* exclude kotlin test from actions...

modmuss50 committed:
Fix typo

modmuss50 committed:
Update to run tests against gradle 7.0 release

modmuss50 committed:
Update mixin-compile-extensions

IotaBread committed:
Add loom development dependencies to test runtime classpath (#389)

modmuss50 committed:
Update MCE

modmuss50 committed:
Merge branch 'dev/0.7' of github.com:FabricMC/fabric-loom into dev/0.7

modmuss50 committed:
Revert "Update MCE"

This reverts commit b85659640d74b534af9c391fffaf6cd62c501292.

modmuss50 committed:
Update MCE to 0.4.3

modmuss50 committed:
Merge branch 'dev/0.7' into future/dev

# Conflicts:
#	src/test/groovy/net/fabricmc/loom/test/util/ProjectTestTrait.groovy

modmuss50 committed:
Update to release 7.0 gradle

modmuss50 committed:
Fix actions

modmuss50 committed:
Make remapped configurations lazy

modmuss50 committed:
Fix gradle 8 deprecation warning

modmuss50 committed:
Fix typo

modmuss50 committed:
Make all loom added configurations lazy

modmuss50 committed:
Java 16

modmuss50 committed:
Enable `ShowCodeDetailsInExceptionMessages` (Better null pointers) for all run configs

modmuss50 committed:
Start 0.8 dev cycle

modmuss50 committed:
Update stitch

modmuss50 committed:
Cleanup merged jar on failure

modmuss50 committed:
Add a simple java 16 project test

modmuss50 committed:
Catch all when merging jars

modmuss50 committed:
General cleanup by making use of java 16 features (#397)

* General cleanup by making use of java 16 features

* use jackson-databind in place of gson when reading to a record

* Fixes

* cleanup

* dep updates

* Replace commons IOUtils usage with native java

* Update fernflower

modmuss50 committed:
Make the run config api stable

modmuss50 committed:
Update MCE to 0.4.4

modmuss50 committed:
Add SignedProjectTest #404

Logiquo committed:
add support for new dependencyResolutionManagement (#400)

* move repo declartions

- Move repository declartions in MavenConfiguration.java to LoomRepositoryPlugin.java

* move repo declartions

- Move repository declartions in MinecraftMappedProvider.java to LoomRepositoryPlugin.java

* move repo declartions

- Move repository declarations in MinecraftProcessedProvider.java to LoomRepositoryPlugin.java

* do not add repositories if dependencyResolutionManagement is used

* Simplify the change on LoomGradlePlugin

- this is the suggestion from liach

* change name to follow fabric naming convension

- change getProjectUUID to getProjectUuid
- change PROJECT_MAPPED_CLASSIFIER to projectMappedClassifier

* remove MavenConfiguration.java

- the file currently do nothing.

* clean-up for all `instanceof` clause

* add DependencyResolutionManagementTest

* code cleanup

* Update src/test/resources/projects/dependencyResolutionManagement/projmap/src/main/resources/modid.accesswidener

* change project uuid to project full name

Co-authored-by: modmuss50 <modmuss50@gmail.com>

magneticflux- committed:
Add classifier to "fake" mods' id if they have a classifier specified (#394)

* Remove unused codepath (`RemapJarTask` is an `AbstractArchiveTask`)

* Remove incorrect comment

Comment originally added in 02e9616622abed6b266ced329b86d419ac31ee88

* Add classifier to `DependencyInfo`

* Add classifier to generated mod ID

* Clean up code structure from review

Co-authored-by: haykam821 <haykam821@users.noreply.github.com>

* Inline `file` local variable

Co-authored-by: haykam821 <haykam821@users.noreply.github.com>

Co-authored-by: haykam821 <haykam821@users.noreply.github.com>

modmuss50 committed:
Improve error messages when using outdated Java or Gradle.

modmuss50 committed:
Improve the outdated Java error message when syncing in idea.

modmuss50 committed:
Suppress "Inconsistent inner class entries" error when decompiling

modmuss50 committed:
Fix local ivy layout not respecting classifiers. Fixes #407

modmuss50 committed:
Fix MergedNestedJarProvider not calling the child's prepare

modmuss50 committed:
Update mercury (#408)

modmuss50 committed:
Revert "Update mercury (#408)"

This reverts commit 50d6fd298d7ac450dea1b8475c43e73255e674b5.

modmuss50 committed:
Run tests on windows

modmuss50 committed:
Start 0.9 dev cycle

modmuss50 committed:
Always fork the java compiler

modmuss50 committed:
Fix new gradle 8.0 deprecation added in 7.2 (#416)

* Fix new gradle 8.0 deprecation added in 7.2

* Handle bootstrap tests a bit better

* Fix 0.9 wrapper tests

haykam821 committed:
Use tabs for indentation in the Fernflower decompiler (#383)

Co-authored-by: modmuss50 <modmuss50@gmail.com>

modmuss50 committed:
Update tiny remapper.

modmuss50 committed:
0.4.1 tiny remapper

modmuss50 committed:
Update Tiny Remapper to 0.4.2

modmuss50 committed:
Basic layered mappings with @ParchmentMC support (#413)

modmuss50 committed:
Ensure deps are reproducible. This prevents them from being updated without us knowing.

modmuss50 committed:
Patch all eclipse deps to use a strict version

shedaniel committed:
Fix typo "officalMojangMappings" (#420)

Signed-off-by: shedaniel <daniel@shedaniel.me>

Player3324 commented:
Sorry for the late reply, I don't usually work on Loom..

IMO we should have enum extension support, but I'd rather leave the enums intact and use a technique similar to what modmuss described instead of breaking them open. This reduces the risk of java and jit compiler problems (including later java versions) since it can generate the exact same bytecode, gives better guarantees about the usual enum invariants by keeping them sealed and doesn't expose enum implementation details to mods where changes in Javac's compilation strategy would introduce breakage that a simple loader update could fix otherwise.

The access widener approach sounds quite suitable, they have the same challenges with making java/ides and the runtime happy. Maybe it should even end up in a common library with AW. Nothing should btw rely on loader's current "make everything public" hack.

nnym committed:
pull upstream

nnym committed:
suffix version by "-enum"

shartte commented:
How does this interact with serializing enums between server and client? I.e. using the read/write enum methods on packet buffers?

Wouldn't this mean that server/client need the same enum mods and also in the same order?

Ueaj-Kerman commented:
I think you could circumvent the order question (eg. by passing an identifier, and sorting them, or simply sorting by name), but you'd still need server/client. Though that begs the question, isn't that always the case? Like if the server adds an enum that the client doesn't have, and it's being networked, well then it will surely fail on the client right?
